### DISCUSSION TITLE
Getting started

### DISCUSSION THREAD

**Discussion by yuppox** [rsta2/circle#452] [General]
Hi,

I tried building for qemu. What do I need to run one of the samples? What should the command line be? I want to emulate a RPi 3.

**Comment by yuppox** [comment]
My command line is `qemu-system-aarch64.exe -M raspi3b -kernel kernel8-32.img` (using sample 41) but QEMU is just a blank window.

**Reply by rsta2** [reply]
You have to build the sample program for the AArch64 mode (*kernel8.img*), not for AArch32 (*kernel8-32.img*), if you want to run it in *qemu-system-aarch64.exe*. You have to set `AARCH=64` in *Config.mk* before build for that purpose. You need a 64-bit toolchain. Don't forget to clean the whole project using `makeall clean` from project root and `make clean` and the sample's directory after changing the configuration in *Config.mk*.

*sample/41-screenanimations* is not perfect for a first screen test, because it uses display double buffering by default, which is not supported by QEMU. Please try *sample/03-screentext* for a first test.

**Reply by yuppox** [reply]
Thank you. I got it working. It was confusing since the 32-bit build chain worked with the real hardware.

One other question - how can I get my keyboard to work with qemu? So far, I can't get example 08 to work. I'm using the `-usb -device usb-kbd` option.

When I start qemu, I see this:
```
00:00:00.55 timer: SpeedFactor is 1.81
00:00:00.56 dwhci: Unknown vendor 0x4F54294A
```

And nothing happens after that.

**Reply by yuppox** [reply]
I patched that vendor ID in qemu and that seems to have fixed the problem.

**Reply by rsta2** [reply]
QEMU normally needs [this patch](https://github.com/smuehlst/qemu/commit/62b39b245b1909c370a365aa2e93fd44830240d2) to work with the Circle USB driver and with networking.

**Reply by yuppox** [reply]
Right, that's what I applied - the second part.

**Comment by yuppox** [comment]
What I don't understand is that I think a 32-bit build works on real hardware. Why won't it work in qemu?

**Reply by yuppox** [reply]
Apparently, this is simply a limitation of QEMU. 👎

**Reply by rsta2** [reply]
There is a version of QEMU (qemu-system-arm), which runs Raspberry Pi 1, 2 and Zero 32-bit binaries.

**Reply by yuppox** [reply]
Thanks. I tried qemu-system-arm. It doesn't support raspi3b, which is kind of annoying but more importantly, it doesn't support the GPU.

**Comment by yuppox** [comment]
> _sample/41-screenanimations_ is not perfect for a first screen test, because it uses display double buffering by default, which is not supported by QEMU. Please try _sample/03-screentext_ for a first test.

I can't find a good emulator for the Pi. QEMU doesn't seem to support the GPU at all and the framebuffer is not very good (no double buffer, as you noted). It seems like we need a simulator for graphics.

**Reply by rsta2** [reply]
When you specify a `FALSE` as third parameter to the constructor of `C2DGraphics`, you can use it without double-buffering.
